proposed/non-primitive-constant-data.md

```diff
@@ -0,0 +1,163 @@
+
+```csharp
+struct NonAlignedStruct
+{
+    byte b;
+    double d;
+}
+```
+
+### RuntimeHelpers apis
+
+```csharp
+class RuntimeHelpers
+{
+    // Fundamental new api capability
+    // Behavior if the values pointed at by inData changes over time is undefined
+    // This function is expected to be implemented as a compiler intrinsic with the behavior of the c# written below. The proposal
+    // expects that implementors will make this work for arbitrary calls, not just as a jit intrinsic, but that's possibly not completely necessary.
+    ReadOnlySpan<TOutput> LoadConstantData<TOutput, TInput>(ref TInput inData, int count) where TOutput:struct where TInput:struct
+    {
+        if (!VerifyThatTypeIsValidForConstantRepresentation(typeof(TOutput)))
+            throw new InvalidProgramException();
+
+        if (count < 0)
+            throw new InvalidProgramException();
+
+        if (checked(GetSizeOfConstantRepresentation(typeof(TOutput)) * count) > sizeof(TInput))
+            throw new InvalidProgramException();
+
+        if (HasGCPointers(typeof(TInput))
+            throw new InvalidProgramException();
+
+        if (!VerifyInDataPointsInsideOfSomeLoadedManagedAssembly(ref inData))
+            throw new InvalidProgramException();
+
+        if (IsAlignedForTOutput(ref inData, typeof(TOutput)) && WellKnownTypeLayoutMatchesPlatformLayout(typeof(TOutput)))
+        {
+            return new ReadOnlySpan<TOutput>(Unsafe.As<TOutput>(inData), count);
+        }
+        else
+        {
+            // Convert inData pointer to raw pointer, and use it as an entry in a hashtable to store converted constant data
+            // The converted constant data will be constructed in some fashion like...
+            IntPTr inDataPtr = (IntPtr)Unsafe.AsPtr(inData);
+            lock(hashtable)
+            {
+                TOutput [] data;
+                if (!hashtable.TryGetValue(inDataPtr, out data) || data.Length < count)
+                {
+                    data = new TOutput[count];
+                    // Read data from inData into count, handling the type layout transition
+                    hashtable[inDataPtr] = data;
+                }
+                return data.AsSpan();
+            }
+        }
+    }
+
+    // Should be treated as an intrinsic by the compiler. Should allow more efficient encoding of single constants in an IL stream
+    TOutput LoadIndividualConstant<TOutput, TInput>(ref TInput inData) where TOutput:struct where TInput:struct
+    {
+        return LoadConstantData(ref inData, 1)[0];
+    }
+
+    // General replacement for ArrayInitialize pattern, also handles non-primitive constants.
```

Comment:
Alternative design is to teach the existing InitializeArray API to support more types, and not introduce new API.

Reply:
InitializeArray requires that there is an array instance. Span-based API feels a bit more flexible.

It might be nice if InitializeArray works with generalized const-able types, but having span based APIs it is not strictly necessary. Compiler couild just emit `ToArray` or something.

Reply:
Sure, its straightforward to enable InitializeArray API to support more types. I generally prefer to avoid extending existing apis, but that would be reasonable to do.